### PR TITLE
Docs: Add SPL Token exchange integration

### LINKED ISSUE #12281: Add an section for SPL Token integration to the exchange integration guide
Considerations:
1. The `spl-token` program is distributed as a source crate only.  We might need to add binary release artifacts, perhaps include spl-token in the standard Solana release artifacts?
2. The per-user deposit address will require the exchange to establish a token account (ie. `spl-token create-account`).  This requires a pinch of Sol per user for rent, and is not quite as convenient as the native token deposit workflow.   The user will need to "register their intent" in the Exchange UI.  There may be other alternatives such as using the SPL Token delegate facility (but sollet.io doesn't do delegates yet), "memo field" for transfers into a common exchange SPL Token deposit address, etc.
3. For withdrawals, some exchanges use `solana transfer --no-wait` to avoid blocking their withdrawal pipeline.  `spl-token transfer` may be fast enough since it only waits for single confirmation, but if not then we'll need a `--no-wait` option here as well.
4. Highlight the mint-level freeze authority to Exchanges.  When they integrate an SPL Token with `Some` freeze authority, an external party has the power to freeze tokens under their custody.   This is something they need to be aware of

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

No documentation to assist exchanges in integrating SPL Token

#### Summary of Changes

Add some

Fixes #12281

~TODO~

- [ ] ~Add `--verbose` flag to `spl-token balance` for also printing the mint~
- [ ] ~Recommend checking withdraw accounts with ☝️~

I'm going to punt the TODO to https://github.com/solana-labs/solana-program-library/issues/467

### CODE REVIEW

**Comment by t-nelson** [comment]
Do we want to mention the existing JS (no docs published yet) or RPC (explicitly tagged unstable) at this point?

**Comment by mvines** [comment]
> Do we want to mention the existing JS (no docs published yet) or RPC (explicitly tagged unstable) at this point?

Na, just another place to update when that's no longer the case IMO.

File: docs/src/integrations/exchange.md [hunk 1]
```diff
@@ -488,10 +488,121 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+✅ Approved
+Signature: 3R6tsog17QM8KfzbcbdP4aoMfwgo6hBggJDVy7dZPVmH2xbCWjEj31JKD53NzMrf25ChFjY7Uv2dfCDq4mGFFyAj
+```
+
+### Other Considerations
+
+#### Freeze Authority
+For regulatory compliance reasons, an SPL Token mint may impose "Freeze Authority"
+over all account created for it.  This allows the issuing entity to [freeze](https://spl.solana.com/token#freezing-accounts)
+the assets in a given account at will, rendering the account unusable until thawed.
```
Comment (mvines) [inline docs/src/integrations/exchange.md:599]: Let's mention that this is an optional feature that may be permanently disabled for any given mint, and then explain how to check that in the explorer.

Which brings us to a new explorer feature!    When viewing a Token mint ([example](https://explorer.solana.com/address/TESTpKgj42ya3st2SQTKiANjTBmncQSCqLAZGcSPLGM)) in the explorer, we don't display the Freeze authority at all.  

We could add `spl-token` support for this as well but I feel that if the explorer shows it then we're 99% of the way there and no need to really duplicate the feature in the cli

File: docs/src/integrations/exchange.md [hunk 2]
```diff
@@ -488,10 +488,121 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+
+1. SPL Token accounts are not implicitly created, so must be created explicitly
+before an SPL Token balance can be deposited
+1. SPL Token accounts must remain [rent-exempt](https://docs.solana.com/apps/rent#rent-exemption)
+for the duration of their existence and therefore require a small amount of 
+native SOL tokens be deposited at account creation. For SPL Token v2 accounts,
+this amount is 2,039,280 Lamports (0.00203928 SOL).
+
+When a user wishes to deposit a new SPL Token type, a new account will need to be
+created for that user at the given mint.
```
Comment (mvines) [inline docs/src/integrations/exchange.md:542]: I think we should just go all-in on the recommendation that the exchange pre-create deposit accounts in a large batch.    

This brings us to a new `solana-tokens` feature -- `distribute-tokens` (but real tokens, not lamports).  Useful here but it also unlocks SPL Token airdrop campaigns too.

**Comment by mvines** [comment]
Let's take the opportunity to flesh out the "Listening for Deposits" section, for both SOL and tokens.   General recommendation is that the `getConfirmedBlocks` and then `getConfirmedBlock` RPC methods be used to monitor the chain for deposits into the exchange user accounts.   

For SOL, scanning the transaction meta `postBalances` for increases to any deposit account is sufficient.   
For tokens, they'd need to check for spl-token transfer instructions into a deposit address and then invoke `getTokenAccountBalance` on the deposit account to read it's current balance.   This nicely adds a dependency on @jstarry's work to expose CPI instructions in RPC as well, since the exchange wouldn't want to miss those deposits

**Comment by mvines** [comment]
There's probably a small handful of issues that are going to come out of this effort @t-nelson.  Consider creating a Github project so that it's easy to track them all and we can have different folks working on some in parallel potentially.

**Comment by mvines** [comment]
This seems to be the best withdrawal flow: 
User provides their primary wallet address.  Exchange creates a new spl-token account and assigns authority of it to the user's wallet.  This will require wallets like Sollet/Solflare to support merging of spl-token accounts (but that seems generally useful regardless).  

Deposit flow is through a normal spl-token transfer operation

**Comment by t-nelson** [comment]
New project: https://github.com/orgs/solana-labs/projects/2

I've populated it with the stuff I'm aware of so far

**Comment by t-nelson** [comment]
> This seems to be the best withdrawal flow:
> User provides their primary wallet address. Exchange creates a new spl-token account and assigns authority of it to the user's wallet. This will require wallets like Sollet/Solflare to support merging of spl-token accounts (but that seems generally useful regardless).

Is this going to fit into existing withdraw UX flows without additional integration overhead?

**Comment by t-nelson** [comment]
> For SOL, scanning the transaction meta `postBalances` for increases to any deposit account is sufficient.

Wasn't there discussion of extending `postBalances` with SPL Token balance deltas as well?  More of a "wouldn't it be nice" than a "let's do it right now" comment, the suggestion sounds find in the interim

**Comment by mvines** [comment]
> > This seems to be the best withdrawal flow:
> > User provides their primary wallet address. Exchange creates a new spl-token account and assigns authority of it to the user's wallet. This will require wallets like Sollet/Solflare to support merging of spl-token accounts (but that seems generally useful regardless).
> 
> Is this going to fit into existing withdraw UX flows without additional integration overhead?

I think so.  Withdraw flow is pretty simple: user provides address and amount.   In this case the address is for the account authority.  Then rather than running a `solana transfer <address> <amount>` to execute the withdraw, it's some `spl-token ...` commands

**Comment by mvines** [comment]
> > For SOL, scanning the transaction meta `postBalances` for increases to any deposit account is sufficient.
> 
> Wasn't there discussion of extending `postBalances` with SPL Token balance deltas as well? More of a "wouldn't it be nice" than a "let's do it right now" comment, the suggestion sounds fine in the interim

Ah that's probably a better solution.  I think we can scope out both and then decide if we want to skip the interim option entirely

**Comment by t-nelson** [comment]
> I think so. Withdraw flow is pretty simple: user provides address and amount. In this case the address is for the account authority. Then rather than running a solana transfer <address> <amount> to execute the withdraw, it's some spl-token ... commands

Yeah, I guess my concern is that there will be significant education required to communicate the indirection here.  I could see a user providing an actual SPL Token account for w/d, which was created with a throwaway keypair for instance, which would be bad.

> Ah that's probably a better solution. I think we can scope out both and then decide if we want to skip the interim option entirely

I'll add an issue!

File: docs/src/integrations/exchange.md [hunk 3]
```diff
@@ -488,6 +488,145 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+few differences which will be discussed in this section.
+
+### Token Mints
+
+Each *type* of SPL Token is declared by creating a *mint* account.  This account
+stores metadata describing token features like the supply, number of decimals and
+various authorities with control over the mint.  Each SPL Token account references
+its associated mint and may only interact with SPL Tokens of that type.
+
+### Installing the `spl-token` CLI Tool
```
Comment (CriesofCarrots) [inline docs/src/integrations/exchange.md:506]: I find the ordering and heading levels to be a bit weird and hard to follow, especially this fairly lengthy section on installing a tool that I don't yet know the need for. Could we shunt the installation information to somewhere different and just link to it under `## Supporting the SPL Token Standard` above? (Something like `Example commands below utilize the SPL Token CLI toll, which can be installed by follow the steps [here](link)`)

File: docs/src/integrations/exchange.md [hunk 4]
```diff
@@ -488,6 +488,145 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+
+#### Example
+```
+$ solana balance 6VzWGL51jLebvnDifvcuEDec17sK6Wupi4gYhm5RzfkV
+0
+```
+
+### Token Transfers
+
+For SPL Token transfers to succeed a few prerequisite conditions must be met
```
Comment (CriesofCarrots) [inline docs/src/integrations/exchange.md:572]:
```suggestion
For SPL Token transfers to succeed, a few prerequisite conditions must be met:
```

File: docs/src/integrations/exchange.md [hunk 5]
```diff
@@ -488,6 +488,145 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+1. SPL Token accounts are not implicitly created, so must be created explicitly
+before an SPL Token balance can be deposited
+1. SPL Token accounts must remain [rent-exempt](https://docs.solana.com/apps/rent#rent-exemption)
+for the duration of their existence and therefore require a small amount of
+native SOL tokens be deposited at account creation. For SPL Token v2 accounts,
+this amount is 2,039,280 Lamports (0.00203928 SOL).
+
+#### Command Line
+To create an SPL Token account, for the given mint at a random address and owned
+by the funding account's keypair
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/src/integrations/exchange.md:543]: Maybe I should just convert it to a numbered list?  I debated doing that when I first wrote it

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md:543]: Let's see what it looks like!

File: docs/src/integrations/exchange.md [hunk 6]
```diff
@@ -488,6 +488,145 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+or [Transfer2](https://github.com/solana-labs/solana-program-library/blob/096d3d4da51a8f63db5160b126ebc56b26346fc8/token/program/src/instruction.rs#L252)
+instructions referencing user accounts, then querying the
+[token account balance](https://docs.solana.com/apps/jsonrpc-api#gettokenaccountbalance)
+updates.
+
+[Considerations](https://github.com/solana-labs/solana/issues/12318) are being
+made to exend the `preBalance` and `postBalance` transaction status metadata
+fields to include SPL Token balance transfers.
+
+### Withdraws
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/src/integrations/exchange.md:612]: I think this is one is the pharmacological condition 😅

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md:612]: Lol, maybe, but I don't think "withdraw" is a noun in any usage... ;)
I actually don't think "withdraw" is an adjective in any context either, so I think you should also change the "withdraw transfer" that I didn't comment on yet.

File: docs/src/integrations/exchange.md (hunk 3 above)
Comment (t-nelson) [reply docs/src/integrations/exchange.md:506]: Yeah I want to clean up https://spl.solana.com/token then link there from here.  It's not ready yet though, so I figured we dupe here to get integration unblocked.

Would an introductory sentence here like, "SPL Token accounts are managed with the `spl-token` command line tool"

File: docs/src/integrations/exchange.md [hunk 7]
```diff
@@ -488,6 +488,145 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+  Sender: 6B199xxzw3PkAm25hGJpjj3Wj3WNYNHzDAnt1tEqg5BN
+  Recipient: 6VzWGL51jLebvnDifvcuEDec17sK6Wupi4gYhm5RzfkV
+✅ Approved
+Signature: 3R6tsog17QM8KfzbcbdP4aoMfwgo6hBggJDVy7dZPVmH2xbCWjEj31JKD53NzMrf25ChFjY7Uv2dfCDq4mGFFyAj
+```
+
+### Deposits
+Since each `(user, mint)` pair requires a separate account on chain, it is
+recommended that batches of token accounts be created in advance and assigned
+to users as requested. These accounts should all be owned by exchange-controlled
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/src/integrations/exchange.md:597]: Yeah sorry, I fall to it unconsciously 😕

File: docs/src/integrations/exchange.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/src/integrations/exchange.md]: https://github.com/solana-labs/solana/pull/12303/commits/939faafa6797de264ba9edb3234f4ceaac628c75

File: docs/src/integrations/exchange.md (hunk 3 above)
Comment (t-nelson) [reply docs/src/integrations/exchange.md:506]: https://github.com/solana-labs/solana/pull/12303/commits/5141cad6244ecec0940cfd6f3abdd893ec80309e

File: docs/src/integrations/exchange.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/src/integrations/exchange.md]: https://github.com/solana-labs/solana/pull/12303/commits/8b8d0b524d204d3ea39eef4dabdf6e12c809e9bb

File: docs/src/integrations/exchange.md [hunk 8]
```diff
@@ -488,6 +488,149 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+[token account balance](https://docs.solana.com/apps/jsonrpc-api#gettokenaccountbalance)
+updates.
+
+[Considerations](https://github.com/solana-labs/solana/issues/12318) are being
+made to exend the `preBalance` and `postBalance` transaction status metadata
+fields to include SPL Token balance transfers.
+
+### Withdrawing
+The withdraw address a user provides must point to an initialized SPL Token account
+of the correct mint. Before executing a withdraw [transfer](#token-transfers),
```
Comment (CriesofCarrots) [inline docs/src/integrations/exchange.md:618]:
```suggestion
of the correct mint. Before executing a withdrawal [transfer](#token-transfers),
```

File: docs/src/integrations/exchange.md (hunk 3 above)
Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md:506]: All right, works for now

File: docs/src/integrations/exchange.md (hunk 8 above)
Comment (t-nelson) [reply docs/src/integrations/exchange.md:618]: No pass on this one since "transfer" is the noun? :swea

File: docs/src/integrations/exchange.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md]: Yep, better.

File: docs/src/integrations/exchange.md (hunk 8 above)
Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md:618]: https://github.com/solana-labs/solana/pull/12303#discussion_r491096121

Comment (CriesofCarrots) [reply docs/src/integrations/exchange.md:618]: It's a bummer that withdrawal has multiple meanings, but it is in fact the correct word for this action. Let's just make sure not to use the phrase "going through withdrawals" in the context of exchanges 😆

Comment (t-nelson) [reply docs/src/integrations/exchange.md]: https://github.com/solana-labs/solana/pull/12303/commits/8a8c061fdf046a965e39970765978ea5335bb918 🎉

File: docs/src/integrations/exchange.md [hunk 9]
```diff
@@ -488,6 +488,149 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+Monitoring for deposit transactions should follow the [block polling](#poll-for-blocks)
+method described above. Each new block should be scanned for successful transactions
+issuing SPL Token [Transfer](https://github.com/solana-labs/solana-program-library/blob/096d3d4da51a8f63db5160b126ebc56b26346fc8/token/program/src/instruction.rs#L92)
+or [Transfer2](https://github.com/solana-labs/solana-program-library/blob/096d3d4da51a8f63db5160b126ebc56b26346fc8/token/program/src/instruction.rs#L252)
+instructions referencing user accounts, then querying the
+[token account balance](https://docs.solana.com/apps/jsonrpc-api#gettokenaccountbalance)
+updates.
+
+[Considerations](https://github.com/solana-labs/solana/issues/12318) are being
+made to exend the `preBalance` and `postBalance` transaction status metadata
```
Comment (CriesofCarrots) [inline docs/src/integrations/exchange.md:613]:
```suggestion
made to extend the `preBalance` and `postBalance` transaction status metadata
```

**Review by CriesofCarrots — APPROVED** [review]
One typo, then lgtm!

File: docs/src/integrations/exchange.md (hunk 9 above)
Comment (t-nelson) [reply docs/src/integrations/exchange.md:613]: Oops!

**Comment by t-nelson** [comment]
@mvines does this cover our bases now?

File: docs/src/integrations/exchange.md [hunk 10]
```diff
@@ -488,6 +488,149 @@ You can also doublecheck whether a particular blockhash is still valid by sendin
...
+
+SPL Token accounts carry additional requirements that native System Program
+accounts do not:
+
+1. SPL Token accounts are not implicitly created, so must be created explicitly
+before an SPL Token balance can be deposited
+1. SPL Token accounts must remain [rent-exempt](https://docs.solana.com/apps/rent#rent-exemption)
+for the duration of their existence and therefore require a small amount of
+native SOL tokens be deposited at account creation. For SPL Token v2 accounts,
+this amount is 0.00203928 SOL (2,039,280 lamports).
```
Comment (mvines) [inline docs/src/integrations/exchange.md:543]:
```suggestion
this amount is 0.00203928 SOL.
```
(nit: humans use SOL, machines use lamports)

**Review by mvines — APPROVED** [review]

**Comment by mvines** [comment]
> @mvines does this cover our bases now?

Yep! I think this is a great start.  There's a number of improvements of course but that's all incremental work
